diff --git a/libavformat/mpegts.c b/libavformat/mpegts.c
index 92baca61a4..9087639c46 100644
--- a/libavformat/mpegts.c
+++ b/libavformat/mpegts.c
@@ -106,6 +106,8 @@ struct Program {

     /** have we found pmt for this program */
     int pmt_found;
+
+    int cur_eit_ver;
 };

 struct MpegTSContext {
@@ -280,6 +282,7 @@ static void clear_program(MpegTSContext *ts, unsigned int programid)
         if (ts->prg[i].id == programid) {
             ts->prg[i].nb_pids = 0;
             ts->prg[i].pmt_found = 0;
+            ts->prg[i].cur_eit_ver = -1;
         }
 }

@@ -300,6 +303,7 @@ static void add_pat_entry(MpegTSContext *ts, unsigned int programid)
     p->id = programid;
     p->nb_pids = 0;
     p->pmt_found = 0;
+    p->cur_eit_ver = -1;
     ts->nb_prg++;
 }

@@ -2294,6 +2298,109 @@ static void sdt_cb(MpegTSFilter *filter, const uint8_t *section, int section_len
     }
 }

+static void eit_cb(MpegTSFilter *filter, const uint8_t *section, int section_len)
+{
+    MpegTSContext *ts = filter->u.section_filter.opaque;
+    AVFormatContext *ac = ts->stream;
+    SectionHeader h1, *h = &h1;
+    const uint8_t *p, *p_end;
+    AVProgram *program;
+    int i, dlen;
+
+    p_end = section + section_len - 4;
+    p     = section;
+    if (parse_section_header(h, &p, p_end) < 0)
+        return;
+    if (h->tid != EIT_TID || h->sec_num != 0)
+        return;
+    if (ts->skip_changes)
+        return;
+
+    program = NULL;
+    for (i = 0; i < ac->nb_programs; i++)
+        if (ac->programs[i]->id == h->id)
+            program = ac->programs[i];
+    if (!program || program->nb_stream_indexes <= 0)
+        return;
+
+    {
+        struct Program *p = get_program(ts, h->id);
+
+        if (!p || p->cur_eit_ver == h->version)
+            return;
+        p->cur_eit_ver = h->version;
+    }
+
+    av_log(ts->stream, AV_LOG_DEBUG, "EIT:\n");
+    hex_dump_debug(ts->stream, section, section_len);
+
+    if (p + 6 + 12 > p_end)
+        return;
+    p += 6 + 10;
+    dlen = get16(&p, p_end) & 0x0fff;
+    if (p + dlen > p_end)
+        return;
+    p_end = p + dlen;
+    for (;;) {
+        int desc_tag, desc_len;
+
+        desc_tag = get8(&p, p_end);
+        if (desc_tag < 0)
+            break;
+        desc_len = get8(&p, p_end);
+        if (desc_len < 0 || p + desc_len > p_end)
+            break;
+
+        av_log(ts->stream, AV_LOG_DEBUG, "tag: 0x%02x len=%d\n",
+               desc_tag, desc_len);
+        switch (desc_tag) {
+        case 0xc4:  /* audio component descriptor */
+            {
+                int comp_tag;
+                AVStream *stream;
+                int is_dmono, is_multi;
+                char lang[4];
+
+                comp_tag = p[2];
+
+                for (i = 0; i < program->nb_stream_indexes; i++) {
+                    int idx = program->stream_index[i];
+
+                    if (idx < 0 || idx > ac->nb_streams)
+                        continue;
+
+                    stream = ac->streams[idx];
+                    if (stream && stream->stream_identifier == comp_tag + 1)
+                        break;
+                }
+                if (i == program->nb_stream_indexes)
+                    break;
+
+                is_dmono = (p[1] == 0x02);
+                is_multi = !!(p[5] & 0x80);
+                memcpy(lang, p + 6, 3);
+                lang[3] = '\0';
+                av_dict_set(&stream->metadata, "language", lang, 0);
+                av_dict_set_int(&stream->metadata, "isdmono", is_dmono, 0);
+
+                av_log(ts->stream, AV_LOG_VERBOSE,
+                       "prog:%d audio[pid: 0x%04x tag: 0x%02x](%s) dmono:%d\n",
+                       program->id, stream->id, comp_tag, lang, is_dmono);
+
+                if (is_multi) {
+                    memcpy(lang, p + 9, 3);
+                    lang[3] = '\0';
+                    av_dict_set(&stream->metadata, "language2", lang, 0);
+                }
+            }
+            break;
+        default:
+            break;
+        }
+        p += desc_len;
+    }
+}
+
 static int parse_pcr(int64_t *ppcr_high, int *ppcr_low,
                      const uint8_t *packet);

@@ -2700,6 +2807,7 @@ static int mpegts_read_header(AVFormatContext *s)
         mpegts_open_section_filter(ts, SDT_PID, sdt_cb, ts, 1);

         mpegts_open_section_filter(ts, PAT_PID, pat_cb, ts, 1);
+        mpegts_open_section_filter(ts, EIT_PID, eit_cb, ts, 1);

         handle_packets(ts, probesize / ts->raw_packet_size);
         /* if could not find service, enable auto_guess */
@@ -2956,6 +3064,7 @@ MpegTSContext *avpriv_mpegts_parse_open(AVFormatContext *s)
     ts->auto_guess = 1;
     mpegts_open_section_filter(ts, SDT_PID, sdt_cb, ts, 1);
     mpegts_open_section_filter(ts, PAT_PID, pat_cb, ts, 1);
+    mpegts_open_section_filter(ts, EIT_PID, eit_cb, ts, 1);

     return ts;
 }
diff --git a/libavformat/mpegts.h b/libavformat/mpegts.h
index 272e2be4f7..acad47f674 100644
--- a/libavformat/mpegts.h
+++ b/libavformat/mpegts.h
@@ -35,12 +35,14 @@
 /* pids */
 #define PAT_PID                 0x0000
 #define SDT_PID                 0x0011
+#define EIT_PID                 0x0012

 /* table ids */
 #define PAT_TID   0x00
 #define PMT_TID   0x02
 #define M4OD_TID  0x05
 #define SDT_TID   0x42
+#define EIT_TID   0x4e

 #define STREAM_TYPE_VIDEO_MPEG1     0x01
 #define STREAM_TYPE_VIDEO_MPEG2     0x02
